File: 46全排列/src/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        if(nums.length == 0){
            return new ArrayList<>();
        }

        List<List<Integer>> ans = new ArrayList<>();

        if(nums.length == 1){
            List<Integer> tmp = new ArrayList<>();
            tmp.add(nums[0]);
            ans.add(tmp);
            return ans;
        }

        for(int i=0; i<nums.length; i++){
            int tmp = nums[i];
            nums[i] = nums[0];
            nums[0] = tmp;
            int[] nextNums = Arrays.copyOfRange(nums,1,nums.length);
            List<List<Integer>> nextAns = permute(nextNums);
            for(List<Integer> list: nextAns){
                list.addFirst(nums[0]);
            }
            ans.addAll(nextAns);
            nums[0] = nums[i];
            nums[i] = tmp;
        }

        return ans;
    }
}

File: 46全排列/src/Solution.cpp
//
// Created by Frank on 2025/10/16.
//
#include <vector>
using namespace std;
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;

        if (nums.size() == 1) {
            ans.push_back(nums);
            return ans;
        }
        for(int i=0; i<nums.size(); i++) {
            int tmp = nums[i];
            nums[i] = nums[0];
            nums[0] = tmp;
            vector<int> next_nums (nums.begin() + 1, nums.end());
            vector<vector<int>> next_nums_permute = permute(next_nums);
            for (auto v : next_nums_permute) {
                v.insert(v.begin(), nums[0]);
                ans.push_back(v);
            }
            nums[0] = nums[i];
            nums[i] = tmp;
        }
        return ans;
    }
};